tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py

```diff
@@ -0,0 +1,321 @@
+
+
+@pytest.mark.performance
+def test_team_matchmaker_performance(player_factory, bench, caplog):
+    # Disable debug logging for performance
+    caplog.set_level(logging.INFO)
+    num_searches = 200
+    matchmaker = TeamMatchMaker()
+
+    searches = [Search([player_factory(1500, 500, ladder_games=1)]) for _ in range(num_searches)]
+
+    with bench:
+        matchmaker.find(searches, 4)
+
+    assert bench.elapsed() < 0.5
+
+
+def test_team_matchmaker_algorithm(player_factory):
+    matchmaker = TeamMatchMaker()
+    s = make_searches(
+        [1251, 1116, 1038, 1332, 1271, 1142, 1045, 1347, 1359, 1348, 1227, 1118, 1058, 1338, 1271, 1137, 1025],
+        player_factory)
+    c1 = CombinedSearch(*[s.pop(), s.pop()])
+    c2 = CombinedSearch(*[s.pop(), s.pop()])
+    c3 = CombinedSearch(*[s.pop(), s.pop()])
+    c4 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    s.append(c1)
+    s.append(c2)
+    s.append(c3)
+    s.append(c4)
+
+    matches, unmatched = matchmaker.find(s, 4)
+
+    assert set(matches[0][0].get_original_searches()) == {c1, s[2], s[5]}
+    assert set(matches[0][1].get_original_searches()) == {c3, s[1], s[6]}
+    assert set(matches[1][0].get_original_searches()) == {c4, s[4]}
+    assert set(matches[1][1].get_original_searches()) == {c2, s[0], s[3]}
+    assert set(unmatched) == {s[7]}
+    for match in matches:
+        assert matchmaker.assign_game_quality(match).quality > config.MINIMUM_GAME_QUALITY
+
+
+def test_team_matchmaker_algorithm_2(player_factory):
+    matchmaker = TeamMatchMaker()
+    s = make_searches(
+        [227, 1531, 1628, 1722, 1415, 1146, 937, 1028, 1315, 1236, 1125, 1252, 1185, 1333, 1263, 1184, 1037],
+        player_factory)
+    c1 = CombinedSearch(*[s.pop(), s.pop()])
+    c2 = CombinedSearch(*[s.pop(), s.pop()])
+    c3 = CombinedSearch(*[s.pop(), s.pop()])
+    c4 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    s.append(c1)
+    s.append(c2)
+    s.append(c3)
+    s.append(c4)
+
+    matches, unmatched = matchmaker.find(s, 4)
+
+    assert set(matches[0][0].get_original_searches()) == {c4, s[4]}
+    assert set(matches[0][1].get_original_searches()) == {c2, c3}
+    assert set(unmatched) == {s[0], s[1], s[2], s[3], s[5], s[6], s[7], c1}
+    for match in matches:
+        assert matchmaker.assign_game_quality(match).quality > config.MINIMUM_GAME_QUALITY
+
+
+@pytest.mark.slow
+@given(
+    searches=st_searches_list(min_size=4, max_players=8),
+    team_size=st.integers(min_value=2, max_value=4)
+)
+# To hit more cases that don't cause an exception
+# max players should depend on team size and
+# searches should have team_size * 2 total players
+def test_make_even_teams(searches, team_size):
+    matchmaker = TeamMatchMaker()
+    try:
+        teams = matchmaker.make_teams(searches, team_size)
+        assert len(teams[0].players) == team_size
+        assert len(teams[1].players) == team_size
+    except UnevenTeamsException:
+        assert True
```

Comment:
Maybe this would be a good place to use `assume`: https://hypothesis.readthedocs.io/en/latest/details.html#making-assumptions
```suggestion
        assume(False)
```
This should (in theory) make hypothesis waste less time trying examples that are not relevant for the test.

Reply:
This actually increases the time the test uses, because it marks examples as invalid and so it has to generate more examples to find 100 valid ones. I created a new strategy (and assumption) that hopefully generates more suitable data.